usage/src/org/apache/cloudstack/quota/QuotaManagerImpl.java

```diff
@@ -0,0 +1,397 @@
+    @Inject
+    private ConfigurationDao _configDao;
+
+    private TimeZone _usageTimezone;
+    private int _aggregationDuration = 0;
+
+    final static BigDecimal s_hoursInMonth = new BigDecimal(30 * 24);
+    final static BigDecimal s_minutesInMonth = new BigDecimal(30 * 24 * 60);
+    final static BigDecimal s_gb = new BigDecimal(1024 * 1024 * 1024);
+
+    int _pid = 0;
+
+    public QuotaManagerImpl() {
+        super();
+    }
+
+    private void mergeConfigs(Map<String, String> dbParams, Map<String, Object> xmlParams) {
+        for (Map.Entry<String, Object> param : xmlParams.entrySet()) {
+            dbParams.put(param.getKey(), (String) param.getValue());
+        }
+    }
+
+    @Override
+    public boolean configure(String name, Map<String, Object> params) throws ConfigurationException {
+        super.configure(name, params);
+
+        Map<String, String> configs = _configDao.getConfiguration(params);
+
+        if (params != null) {
+            mergeConfigs(configs, params);
+        }
+
+        String aggregationRange = configs.get("usage.stats.job.aggregation.range");
+        String timeZoneStr = configs.get("usage.aggregation.timezone");
+
+        if (timeZoneStr == null) {
+            timeZoneStr = "GMT";
+        }
+        _usageTimezone = TimeZone.getTimeZone(timeZoneStr);
+
+        _aggregationDuration = Integer.parseInt(aggregationRange);
+        if (_aggregationDuration < UsageUtils.USAGE_AGGREGATION_RANGE_MIN) {
+            s_logger.warn("Usage stats job aggregation range is to small, using the minimum value of " + UsageUtils.USAGE_AGGREGATION_RANGE_MIN);
+            _aggregationDuration = UsageUtils.USAGE_AGGREGATION_RANGE_MIN;
+        }
+        s_logger.info("Usage timezone = " + _usageTimezone + " AggregationDuration=" + _aggregationDuration);
+
+        return true;
+    }
+
+    @Override
+    public boolean start() {
+        if (s_logger.isInfoEnabled()) {
+            s_logger.info("Starting Quota Manager");
+        }
+        _pid = Integer.parseInt(System.getProperty("pid"));
+        return true;
+    }
+
+    @Override
+    public boolean stop() {
+        if (s_logger.isInfoEnabled()) {
+            s_logger.info("Stopping Quota Manager");
+        }
+        return true;
+    }
+
+    public List<QuotaUsageVO> aggregatePendingQuotaRecordsForAccount(final AccountVO account, final Pair<List<? extends UsageVO>, Integer> usageRecords) {
+        List<QuotaUsageVO> quotaListForAccount = new ArrayList<>();
+        if (usageRecords == null || usageRecords.first() == null || usageRecords.first().isEmpty()) {
+            return quotaListForAccount;
+        }
+        s_logger.info("Getting pending quota records for account=" + account.getAccountName());
+        for (UsageVO usageRecord: usageRecords.first()) {
+            BigDecimal aggregationRatio = new BigDecimal(_aggregationDuration).divide(s_minutesInMonth, 8, RoundingMode.HALF_EVEN);
+            switch (usageRecord.getUsageType()) {
```

Comment:
@bhaisaab switching on a enumerated type has a number of problems.  First, there is no compiler check to ensure that all enumeration keys are handled.   Any changes to the enumeration's keys require much wider changes throughout code base.  In the case of addition, it is very likely that ``switch`` statements will be missed.  Second, it is very difficult to test these switch statements properly.  It requires setting up elaborate fixtures and mocking.  Finally, these switch blocks can get very long and difficult to follow.  In short, the are a hard to test, leaky abstraction.

When you control the enumeration implementation a better approach is to define an abstract method on the enumeration for the condition passing in any required state and implement it for every value.  Not only does this make the code far more readable (replace a big ``switch`` block with a one line callback that expresses the intent), but it far easier to write a unit test for the enumeration that verifies the operation of each implementation.  Most importantly, the compiler will check that the method is implemented for each value -- ensuring that changes to the structure of the enumeration can be made safely.